operator/pkg/ingress/ingress.go

```diff
@@ -0,0 +1,524 @@
+		})
+	}
+	return networkingv1.IngressStatus{
+		LoadBalancer: v1.LoadBalancerStatus{
+			Ingress: ingresses,
+		},
+	}
+}
+
+func getIngressForStatusUpdate(slimIngress *slim_networkingv1.Ingress, lb slim_corev1.LoadBalancerStatus) *networkingv1.Ingress {
+	slimIngressCopy := slimIngress.DeepCopy()
+	return &networkingv1.Ingress{
+		TypeMeta:   toV1TypeMeta(slimIngressCopy.TypeMeta),
+		ObjectMeta: toV1ObjectMeta(slimIngressCopy.ObjectMeta),
+		Status:     toV1IngressStatus(lb),
+	}
+}
+
+func (ic *IngressController) handleEvent(event interface{}) error {
+	var err error
+	switch ev := event.(type) {
+	case ingressAddedEvent:
+		log.WithField("ingress", ev.ingress.Name).Info("Handling ingress added event")
+		err = ic.handleIngressAddedEvent(ev)
+		break
+	case ingressUpdatedEvent:
+		log.WithField("ingress", ev.newIngress.Name).Info("handling ingress updated event")
+		err = ic.handleIngressUpdatedEvent(ev)
+		break
+	case ingressDeletedEvent:
+		log.WithField("ingress", ev.ingress.Name).Info("Handling ingress deleted event")
+		err = ic.handleIngressDeletedEvent(ev)
+		break
+	case ingressServiceUpdatedEvent:
+		log.WithField("service", ev.ingressService.Name).Info("Handling ingress service updated event")
+		err = ic.handleIngressServiceUpdatedEvent(ev)
+		break
+	default:
+		err = fmt.Errorf("received an unknown event: %s", ev)
+	}
+	return err
+}
+
+func (ic *IngressController) getByKey(key string) (*slim_networkingv1.Ingress, error) {
+	objFromCache, exists, err := ic.ingressStore.GetByKey(key)
+	if err != nil {
+		log.WithError(err).Warn("Ingress cache lookup failed")
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("ingress '%s' not found", key)
+	}
+	ingress, ok := objFromCache.(*slim_networkingv1.Ingress)
+	if !ok {
+		return nil, fmt.Errorf("got an invalid object from ingress cache")
+	}
+	return ingress, nil
+}
+
+func getServiceNameForIngress(ingress *slim_networkingv1.Ingress) string {
+	return ciliumIngressPrefix + ingress.Name
+}
+
+func getIngressKeyForService(service *slim_corev1.Service) string {
+	ingressName := strings.TrimPrefix(service.Name, ciliumIngressPrefix)
+	return fmt.Sprintf("%s/%s", service.Namespace, ingressName)
+}
+
+func (ic *IngressController) createLoadBalancer(ingress *slim_networkingv1.Ingress) error {
+	svc := getServiceForIngress(ingress)
+	svcKey, err := cache.MetaNamespaceKeyFunc(svc)
+	if err != nil {
+		log.Warn("failed to get service key for ingress")
+		return err
+	}
+
+	_, exists, err := ic.serviceManager.getByKey(svcKey)
+	if err != nil {
+		log.WithError(err).Warn("service lookup returned an error")
+		return err
+	}
+	if exists {
+		// Service already exists in the cache. For now assume that it was created by the ingress
+		// controller.
+		log.WithField("service", svcKey).Info("Service already exists. Continuing...")
+		return nil
+	}
+
+	_, err = k8s.Client().CoreV1().Services(ingress.Namespace).Create(context.Background(), svc, metav1.CreateOptions{})
```

Comment:
Out of curiosity, why are we creating Services? :thinking:

Reply:
The goal of this PR is to keep the same design in feature branch without much changes (unless it's blocking).

In my opinion, we will need LB-type service one way or another to route external traffic to cluster, the current design is Ingress owns {Service,Endpoint,CEC} objects. However, there is a discussion to move LB service outside of Ingress controller:

- LB service will be provisioned and managed by user externally.
- Multiple ingress objects can use the  _same_ LB, one benefit is cost saving, the drawback is about handling path conflicts.

This is related to the point mentioned by Michi above

> right now we create one load balancer for each ingress resource. make it so that multiple ingress resources can share a single load balancer.